chore(observability): Adjust log fields and metric tags in enterprise reporting

001wwang opened the pull request:
Closes [OP-398](https://datadoghq.atlassian.net/browse/OP-398)

This PR adjusts internal logs and metrics used for enterprise reporting.
- For metrics
  - Renames the `version` tag to `configuration_version_hash`
  - Adds a new `vector_version` tag
- For logs
  - Renames the top level `version` field (referring to the configuration version hash) to `configuration_version_hash` 
  - Moves all Vector related metadata (configuration_key, configuration_version_hash) under the existing `vector` namespace.

Also, some minor refactoring (renaming, use f-strings).

On the UI side, this requires revising queries that rely on `version` and `configuration_key` for any existing logs and metrics integrations, since these are now renamed/moved.

For example, a new log query would look like

<img width="1268" alt="Screen Shot 2022-05-25 at 3 04 41 PM" src="https://user-images.githubusercontent.com/43912346/170375578-ce02b72e-efd5-4241-af5f-64c7031150ec.png">

### Example

With the following configuration and your own `[enterprise]` section, run Vector.
```
[enterprise]
...

[sources.in]
type = "demo_logs"
format = "json"

[sinks.out]
type = "blackhole"
inputs = ["*"]
```

Navigate to Datadog logs explorer or metrics explorer. For logs, you should see items like the following 

<img width="697" alt="Screen Shot 2022-05-25 at 3 04 11 PM" src="https://user-images.githubusercontent.com/43912346/170375373-43e15abb-b925-4708-94ac-6673b2d6e92d.png">

Note, if you're wondering why `component_*` fields aren't available, it's because these `blackhole` sink logs don't occur in their proper span.

jszwedko commented on src/config/enterprise.rs:
```diff
@@ -375,26 +375,25 @@ fn setup_logs_reporting(
+                "configuration_key": "{configuration_key}",
+                "version_hash": "{version_hash}",
+                "version": "{vector_version}",
+                "arch": "{build_arch}",
+                "os": "{build_os}",
+                "vendor": "{build_vendor}"
```
I wonder if we should add these to the internal metrics too, in the future. I could see users being interested in viewing the metrics by, e.g., OS.

jszwedko reviewed (commented):
Implementation looks good. Would calling `version_hash`, `configuration_hash` make more sense though? That would clearly link it to the configuration and with `configuration_key`. `version_hash` says to me that it has something to do with Vector's version rather than its configuration.

001wwang commented:
> Implementation looks good. Would calling `version_hash`, `configuration_hash` make more sense though? That would clearly link it to the configuration and with `configuration_key`. `version_hash` says to me that it has something to do with Vector's version rather than its configuration.

That's a fair point. @sghall has some good thoughts here on why `version_hash` might fit: https://dd.slack.com/archives/C01UQ5ZPV6K/p1653112974751869?thread_ts=1653077628.133229&cid=C01UQ5ZPV6K. 

Maybe the best option is to be as explicit as possible and title it `config_version_hash`? At least from a UI perspective, "version" is a pretty front-and-center concept, so I think it's worthwhile to have it in there. I'm happy to change it to whatever users would find most intuitive.

leebenson commented:
> Maybe the best option is to be as explicit as possible and title it config_version_hash/configuration_version_hash? At least from a UI perspective, "version" is a pretty front-and-center concept, so I think it's worthwhile to have it in there. I'm happy to change it to whatever users would find most intuitive. cc: @leebenson, @binarylogic

For the avoidance of doubt, I'd prefer we opt `configuration_version_hash`.

I think this satisfies the nomenclature we use within OP, and makes it clearer within Vector that this refers to a config and not some kind of build version related hash.

cc @jszwedko and @sghall for sign off.

leebenson reviewed (changes requested):
Per my [last comment](https://github.com/vectordotdev/vector/pull/12852#issuecomment-1137026328), I think we should standardise the version hash as:

```
configuration_version_hash
```

Between McNulty, Postgres schema, the OP API and Vector, we've managed to use pretty much every variation besides this one 😅 

- config_version
- version_hash
- config_hash
- config_version_hash

Ideally, we could start to use `configuration_version_hash` _everywhere_, so it's clearer that this refers to the same thing.

(The exception I'd make for now is DB schema. I don't want to break anything within McNulty. We can update the schema later when we've fully migrated to the new API.)

I'd like @sghall and @jszwedko to approve these changes first, since this relates to both UI consumption and Vector's reporting in general. If they're in agreement with the tag name, we can then tee up follow-up work in the API to sync.

jszwedko commented:
`configuration_version_hash` sounds reasonable to me. I still think `version` isn't quite the right word for what it is, but I'm ok with it. It's really just a hash of the configuration. If I was picking something, I'd probably go with `configuration_hash`.

001wwang commented:
For posterity, @jszwedko, @sghall, and I had a discussion in Slack. Based on the comments here and the discussion, I've decided to move ahead with `configuration_version_hash`. While it isn't necessarily everyone's first preference, it seems like the best compromise and, at worst, is the most explicit option.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -135,7 +135,7 @@ struct PipelinesAuth<'a> {
 /// Holds the relevant fields for reporting a configuration to Datadog Observability Pipelines.
 struct PipelinesStrFields<'a> {
-    config_version: &'a str,
+    config_version_hash: &'a str,
```
Should this also be the expanded `configuration_version_hash`?

It may seem a bit redundant, but these internal discrepancies can be confusing when multiplied, especially when considering the interface between Vector and the API.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -237,7 +237,7 @@ impl<'a> PipelinesVersionPayload<'a> {
             data: PipelinesData {
                 attributes: PipelinesAttributes {
-                    config_hash: fields.config_version,
+                    config_hash: fields.config_version_hash,
```
Same here. We are sending a `config_version_hash` to a `config_hash` endpoint.

I realise this is a necessity right now given that we're still reporting to McNulty (and we definitely don't want to waste time updating McNulty), but for now `config_version_hash` should be `configuration_version_hash` and `config_hash` (the actual key/value that makes up the payload) can be updated in the API per https://datadoghq.atlassian.net/browse/OP-400

leebenson commented on src/config/enterprise.rs:
```diff
@@ -257,7 +257,7 @@ impl<'a> PipelinesVersionPayload<'a> {
     pub opts: Options,
     pub api_key: String,
-    pub config_version: String,
+    pub config_version_hash: String,
```
Expanded form?

leebenson commented on src/config/enterprise.rs:
```diff
@@ -290,13 +290,13 @@ impl TryFrom<&Config> for EnterpriseMetadata {
-        // Get the configuration version. In DD Pipelines, this is referred to as the 'config hash'.
-        let config_version = value.version.clone().expect("Config should be versioned");
+        // Get the configuration version hash. In DD Pipelines, this is referred to as the 'config hash'.
+        let config_version_hash = value.version.clone().expect("Config should be versioned");
```
Per https://datadoghq.atlassian.net/browse/OP-400, this can later be the expanded form too.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -290,13 +290,13 @@ impl TryFrom<&Config> for EnterpriseMetadata {
             opts,
             api_key,
-            config_version,
+            config_version_hash,
```
Expanded form.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -379,25 +379,25 @@ pub(crate) fn report_on_reload(
 pub(crate) fn attach_enterprise_components(config: &mut Config, metadata: &EnterpriseMetadata) {
     let api_key = metadata.api_key.clone();
-    let config_version = metadata.config_version.clone();
+    let configuration_version_hash = metadata.config_version_hash.clone();
```
A note to sync these to both `configuration_version_hash`

leebenson commented on src/config/enterprise.rs:
```diff
@@ -553,7 +555,7 @@ pub(crate) fn report_configuration(
         let EnterpriseMetadata {
             api_key,
-            config_version,
+            config_version_hash: config_version,
```
Expanded form when relevant.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -570,7 +572,7 @@ pub(crate) fn report_configuration(
         // exploding as func arguments to avoid confusion with multiple &str fields.
         let fields = PipelinesStrFields {
-            config_version: config_version.as_ref(),
+            config_version_hash: config_version.as_ref(),
```
Expanded form when relevant.

leebenson commented on src/config/enterprise.rs:
```diff
@@ -763,7 +765,7 @@ mod test {
     const fn get_pipelines_fields() -> PipelinesStrFields<'static> {
         PipelinesStrFields {
-            config_version: "config_version",
+            config_version_hash: "config_version",
```
Expanded form when relevant.

leebenson reviewed (approved):
Just a few comments to do a final audit and switch to `configuration_version_hash` where relevant.

We're currently in the limbo state of reporting to McNulty but _almost_ being ready to report to our own API endpoint, which gives us full control over the fields we use in the body of a reporting payload. Obviously, we can't make _all_ the changes here to standardise the name just yet until we fully shift over in https://datadoghq.atlassian.net/browse/OP-400.

With that said, it looks like there's a few places where we still refer to this as a "config_version_hash" that we might be able to clean up, and keep just the final payload as the McNulty version for now?

This doesn't need to block the merge. I've updated https://datadoghq.atlassian.net/browse/OP-400 with notes to audit both Vector and the API to sync the version field when we switch over to the final endpoint.

001wwang commented:
> Just a few comments to do a final audit and switch to `configuration_version_hash` where relevant.
> 
> We're currently in the limbo state of reporting to McNulty but _almost_ being ready to report to our own API endpoint, which gives us full control over the fields we use in the body of a reporting payload. Obviously, we can't make _all_ the changes here to standardise the name just yet until we fully shift over in https://datadoghq.atlassian.net/browse/OP-400.
> 
> With that said, it looks like there's a few places where we still refer to this as a "config_version_hash" that we might be able to clean up, and keep just the final payload as the McNulty version for now?
> 
> This doesn't need to block the merge. I've updated https://datadoghq.atlassian.net/browse/OP-400 with notes to audit both Vector and the API to sync the version field when we switch over to the final endpoint.

I took another pass and updated a few straggling `config_version_hash/config_version`'s. I left the `version` field on `Config` as is since that's conceptually correct and seems redundant/overkill to label it `configuration_version_hash`. But, happy to change that too if we'd like.

In the future, on the Vector side, changing the API payload should be as simple as updating the `config_hash` field in `PipelinesAttributes`.

leebenson commented:
Looks good! Thanks for the clean up.
